### PR TITLE
Pippa/flex local strided attn

### PR DESCRIPTION
**Pull request by pippaduckett1**

### CODE REVIEW

File: src/hepattn/models/decoder.py [hunk 1]
```diff
@@ -140,8 +150,8 @@
                 for input_name, task_attn_mask in attn_masks.items():
                     attn_mask[x[f"key_is_{input_name}"].unsqueeze(1).expand_as(attn_mask)] = task_attn_mask.flatten()
 
-            if attn_mask is not None:
-                outputs[f"layer_{layer_index}"]["attn_mask"] = attn_mask
+                if attn_mask is not None:
+                    outputs[f"layer_{layer_index}"]["attn_mask"] = attn_mask
```
[the remark this answers is not in the document]
Comment (pippaduckett1) [reply src/hepattn/models/decoder.py:154]: Left as is because flex implementation will fail if try to log attn_mask

**Review by samvanstroud — COMMENTED** [review]
Looking good thanks @pippaduckett1!

File: tests/models/test_decoder.py [hunk 2]
```diff
@@ -267,6 +269,86 @@ def test_attn_mask_construction(self, decoder, sample_decoder_data):
             assert not attn_mask[0, 1, 3]
             assert not attn_mask[1, 4, 5]
 
+    def test_flex_local_ca_mask_equivalence(self):
```
Comment (samvanstroud) [inline tests/models/test_decoder.py:272]: Could we move these to be standalone tests under `tests/flex/`? Better to avoid couplng to the decoder class here if we just want to test the mask correctness (if we wanted to test the decoder integration then I think a test setup comparing two decoders, one with torch attention and one with flex and running allclose on the outputs would make more sense)

**Review by samvanstroud — COMMENTED** [review]
thanks @pippaduckett1, I think if we simplify the tests and add ignore codes for the linting errors we can merge this :)

File: src/hepattn/flex/local_ca.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        raise ValueError("Window size must be even for strided sliding window")
+
+    stride_val = torch.tensor(stride, device=dev)
+
+    def mask_mod(b, h, q_idx, kv_idx):  # noqa: ARG001
+        # b = 0, h = 0 here
+        q_center = torch.round(q_idx * stride_val)
+        return (kv_idx - q_center).abs() <= window_size // 2
+
+    return create_block_mask(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=dev)
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:17]: I think we might want to return a compiled function, rather than compile this function

```suggestion
def sliding_window_mask_strided(window_size: int, stride: float, q_len: int, kv_len: int, compile: bool = False, device: str | torch.device) -> _mask_mod_signature:
    if window_size % 2 != 0:
        raise ValueError("Window size must be even for strided sliding window")

    stride_val = torch.tensor(stride, device=device)

    def mask_mod(b, h, q_idx, kv_idx):  # noqa: ARG001
        # b = 0, h = 0 here
        q_center = torch.round(q_idx * stride_val)
        return (kv_idx - q_center).abs() <= window_size // 2
    
    block_mask = create_block_mask(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=device)
    
    if compile:
        block_mask = torch.compile(block_mask)
        
    return 
```

File: src/hepattn/flex/local_ca.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+
+@torch.compile(dynamic=True)
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:20]: same as above

File: src/hepattn/flex/local_ca.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    """Exact transpose of a FlexAttention BlockMask by transposing the mask function.
+
+    Args:
+        bm: forward BlockMask (built with Q_LEN=q_tokens, KV_LEN=kv_tokens)
+        q_tokens: original forward Q token length
+        kv_tokens: original forward KV token length
+        device: torch.device to build the transposed mask on (defaults to cpu if None)
+    """
+    orig_mod = bm.mask_mod
+    dev = device if device is not None else getattr(bm, "device", "cpu")
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:51]: does the block mask not always have a device? in which case we probably don't need a dedicated `device` argument

File: src/hepattn/models/decoder.py [hunk 6]
```diff
@@ -154,13 +165,25 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
...
             )
 
             # update the individual input constituent representations
             x = unmerge_inputs(x, input_names)
 
         return x, outputs
 
+    def flex_local_ca_mask(self, q_len: int, kv_len: int, device):
+        # Calculate stride based on the ratio of key length to query length
+        stride = kv_len / q_len
```
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/decoder.py:179]: I agree `round` is better actually! but maybe let's call it here so that `stride` is always an `int`, which is more intuitive. the result should be the same

File: src/hepattn/models/decoder.py [hunk 7]
```diff
@@ -202,16 +221,16 @@ def __init__(
         attn_norm = norm if not hybrid_norm else None
         dense_post_norm = not hybrid_norm
 
-        attn_kwargs = attn_kwargs or {}
+        self.attn_kwargs = attn_kwargs or {}
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:224]: I'd rather not store the `kwargs` as a member, let's just pull out what we need and store that.

File: src/hepattn/models/decoder.py [hunk 8]
```diff
@@ -251,8 +272,10 @@ def forward(
         # Update key/constituent embeddings with the query/object embeddings
         if self.bidirectional_ca:
             if attn_mask is not None:
+                if self.attn_kwargs.get("attn_type", "torch") == "flex":
+                    assert attn_mask_transpose is not None, "attn_mask_transpose must be provided for flex attention"
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:276]: See above, nicer to store `self.attn_type` from the `kwargs` initially and then use it here

File: tests/flex/test_local_ca_flex.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import pytest
+import torch
+from torch.nn.attention.flex_attention import create_mask
+
+from hepattn.flex.local_ca import transpose_blockmask
+from hepattn.models.decoder import MaskFormerDecoder
+from hepattn.utils.local_ca import auto_local_ca_mask
+
+
+@pytest.mark.skipif(os.environ.get("TORCH_COMPILE") == "1", reason="Fails with torch.compile/Inductor due to dynamic shapes")
```
Comment (samvanstroud) [inline tests/flex/test_local_ca_flex.py:12]: Might be important to understand which part of the compilation fails here, does it change if you remove the compile decorators as above?

Can we still test a non-compiled version in the CI?

File: src/hepattn/flex/local_ca.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        return orig_mod(b, h, kv_idx, q_idx)
+
+    return create_block_mask(
+        mask_mod_t,
+        B=None,
+        H=None,
+        Q_LEN=kv_tokens,  # new queries = old KV tokens
+        KV_LEN=q_tokens,  # new keys    = old Q tokens
+        device=dev,
+    )
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:65]: This is a nice trick, I wonder if it's possible to do it with the block mask too, so that we can avoid recreating it? something to follow up on maybe

**Review by samvanstroud — COMMENTED** [review]
Looking really good @pippaduckett1!

File: src/hepattn/models/decoder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pippaduckett1) [reply src/hepattn/models/decoder.py:179]: sorry I was really unclear - what I meant was inside the mask we do `round(q_idx*stride)` as opposed to `q_idx * round(stride)` - but could change this if rounding stride alone would work better

File: src/hepattn/flex/local_ca.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        raise ValueError("Window size must be even for strided sliding window")
+
+    stride_val = torch.tensor(stride, device=device)
+
+    def mask_mod(b, h, q_idx, kv_idx):  # noqa: ARG001
+        # b = 0, h = 0 here
+        q_center = torch.round(q_idx * stride_val)
+        return (kv_idx - q_center).abs() <= window_size // 2
+
+    return create_block_mask(mask_mod, B=0, H=0, Q_LEN=q_len, KV_LEN=kv_len, device=device, _compile=True)
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:22]: `_compile` is deprecated, using torch.compile() is now the recommendation here

File: src/hepattn/flex/local_ca.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    """Exact transpose of a FlexAttention BlockMask by transposing the mask function.
+
+    Args:
+        bm: forward BlockMask (built with Q_LEN=q_tokens, KV_LEN=kv_tokens)
+        q_tokens: original forward Q token length
+        kv_tokens: original forward KV token length
+        device: torch.device to build the transposed mask on (defaults to cpu if None)
+    """
+    orig_mod = bm.mask_mod
+    dev = str(getattr(bm, "device", "cpu"))
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:61]: Why not just `bm.device`?

File: src/hepattn/flex/local_ca.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        return orig_mod(b, h, kv_idx, q_idx)
+
+    return create_block_mask(
+        mask_mod_t,
+        B=0,
+        H=0,
+        Q_LEN=kv_tokens,  # new queries = old KV tokens
+        KV_LEN=q_tokens,  # new keys    = old Q tokens
+        device=dev,
+        _compile=True,
```
Comment (samvanstroud) [inline src/hepattn/flex/local_ca.py:75]: same here

**Review by samvanstroud — COMMENTED** [review]
Thanks for the changes, a few last small comments. 

I'm not sure why the tests are failing - if you can't figure it out easily let's just merge with the test skipped in the ci as it was and figure it out later :)
